benchmarks/vendor_allocator_benchmarks.cpp

```diff
@@ -0,0 +1,183 @@
+//////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2016-20, Lawrence Livermore National Security, LLC and Umpire
+// project contributors. See the COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (MIT)
+//////////////////////////////////////////////////////////////////////////////
+#include "benchmark/benchmark.h"
+
+#include "umpire/Allocator.hpp"
+#include "umpire/ResourceManager.hpp"
+
+#include "umpire/alloc/MallocAllocator.hpp"
+#if defined(UMPIRE_ENABLE_CUDA)
+#include "umpire/alloc/CudaMallocAllocator.hpp"
+#include "umpire/alloc/CudaMallocManagedAllocator.hpp"
+#include "umpire/alloc/CudaPinnedAllocator.hpp"
+#endif
+#if defined(UMPIRE_ENABLE_HIP)
+#include "umpire/alloc/HipMallocAllocator.hpp"
+#include "umpire/alloc/HipMallocManagedAllocator.hpp"
+#include "umpire/alloc/HipPinnedAllocator.hpp"
+#endif
+
+static const int RangeLow{1048576}; //2^20 or ~1MB
+static const int RangeHi{268435456}; //2^28 or ~256MB
+
+/*
+ * Allocate either LARGE (about 12GB), MEDIUM (about 6GB)
+ * or SMALL (about 1GB) for benchmark measurements.
+ */
+//#define LARGE 12000000000
+//#define MEDIUM 6000000000
+#define SMALL 1000000000
+
+class AllocatorBenchmark : public benchmark::Fixture {
+public:
+  using ::benchmark::Fixture::SetUp;
+  using ::benchmark::Fixture::TearDown;
+
+  virtual void* allocate(std::size_t nbytes) = 0;
+  virtual void deallocate(void* ptr) = 0;
+
+  void allocation(benchmark::State& st) {
+    const std::size_t size{static_cast<std::size_t>(st.range(0))};
+    std::size_t i{0};
+    
+    Max_Allocations = setBounds(size);
+    m_allocations.resize(Max_Allocations);
+
+    while (st.KeepRunning()) {
+      if (i == Max_Allocations) {
+        st.PauseTiming();
+        for (std::size_t j{0}; j < Max_Allocations; j++)
+          deallocate(m_allocations[j]);
+        i = 0;
+        st.ResumeTiming();
+      }
+      m_allocations[i++] = allocate(size);
+    }
+
+    // This says that we process with the rate of state.range(0) bytes every iteration:
+    st.counters["BytesProcessed"] = benchmark::Counter(st.range(0), benchmark::Counter::kIsIterationInvariantRate, benchmark::Counter::OneK::kIs1024);
```

Comment:
Do you think we should convert the number to Kb or Mb here?

Reply:
We decided not to do any conversions here. @mcfadden8 any thoughts? (When I divided by 1000 to get MB, the units that google benchmark reported in were confusing anyways, so it was almost more clear to leave it as is)